### DISCUSSION TITLE
Best practices for managing migration scripts?

### DISCUSSION THREAD

**Discussion by bryanrsmith** [planetscale/discussion#69] [Documentation] [answered]
Is there any documentation or recommendations about how best to manage migration scripts in a way that would work well with planetscale long-term? I see that [planetscale's flux app](https://github.com/planetscale/flux) is using `db-migrate`, but I'm unsure of how that tool aligns with schema branches. In particular,

1. Schema changes are frequently accompanied by data migrations. Since schema changes can't be made directly on main, and data changes made on branches don't affect main, what's the best way to perform a schema change in conjunction with a data migration? [The flux app has such a change script here](https://github.com/planetscale/flux/blob/main/migrations/sqls/20210209190815-post-last-viewed-up.sql#L15-L25); how was this migration executed against a production planetscale database?
2. `db-migrate` inserts rows into a `migrations` table in order to track which migrations have been applied. Since these records are not merged into main when the deploy request is accepted, migration history is lost when the branch is discarded and `db-migrate` can't track which migrations have been applied when the next branch is created.

Thanks!

**Comment by jgreet** [comment, marked as answer]
Hi @bryanrsmith, thanks for the detailed questions.

1) I think a source of confusion here is that [planetscale/flux](https://github.com/planetscale/flux) was released before the beta for PlanetScale began and it isn't entirely compatible with our new product yet. Once https://github.com/planetscale/beta/discussions/15 ships we should be able to move it over. <br><br>As you've identified, schema changes and production data transitions will be run in different branches, so a migration containing both should be split into two migrations.


2. We're solving this problem with Automatic Schema Migrations and support for Rails migrations is already live. When enabled on a database's settings page PlanetScale will keep the `schema_migrations` tables in your `main` and development branches updated. Most frameworks, and `db-migrate`, use a similar table and support for additional frameworks is coming soon. <br><br>See also: https://docs.planetscale.com/tutorial/automatic-rails-migrations

**Comment by bryanrsmith** [comment]
Thanks! That's all very helpful. I had overlooked that documentation page because I don't use rails, but it's otherwise just what I was looking for.

**Reply by jgreet** [reply]
@bryanrsmith just wanted to mention that support for other migration tables shipped today. You can check out the feature on the settings page of any database. Updated docs are coming soon.

**Reply by bryanrsmith** [reply]
Perfect! Thanks for the update.

**Reply by bryanrsmith** [reply]
This does not seem to be working for me.

* I have enabled automatic schema migrations.
* I set the framework to `other`, and the table name to `migrations` (the default used by `db-migrate`, which I previously used to set up the initial schema).
* I copied the migration records from my `init` branch that I used to set up the initial schema to the `main` branch, and verified that the data exists on `main`.
* Creating a new branch fails to populate the `migrations` table with any rows, leaving me with an empty table as before.

I also note that when I return to the database settings page, the framework is set to `laravel`. I guess that's because laravel also uses the table name `migrations`, but it's a little confusing because I don't use laravel and had explicitly chose `other`.

Have I missed a step?

**Reply by jgreet** [reply]
We'll dig in on these issues tomorrow. Thanks for letting us know.

**Reply by iheanyi** [reply]
@bryanrsmith Could you try the flow again and report back? If it works, it may have been a transient error, but if it happens again, we'll have to dig into it deeper. I made some changes that I hope fixes this. Thanks in advance.

**Reply by bryanrsmith** [reply]
@iheanyi The settings page now correctly indicates the `Other` migration framework I have selected, but the table still fails to populate when I create a new branch.

Other than a single record in the `migrations` table there is no data in this database yet, so you are welcome to connect to it if it would help your investigation: `99spokes/99spokes`

**Reply by iheanyi** [reply]
@bryanrsmith Thanks for this, we'll have to do more investigating. There may be a bug on our end that's not populating the table, perhaps. I wonder, could you create another database and try it out again please? I want ot make sure it's nothing wrong with that database in particular

**Reply by bryanrsmith** [reply]
@iheanyi Yes, it does appear to be a problem with that specific database. I created a new database and this feature is working as expected there. In the malfunctioning database I've tried toggling the automatic migrations setting off and on and created several branches with no luck.

I'm happy to delete and recreate the database (I'd like to keep the name I chose for this database), but will wait for your word in case you want to investigate further.

(sorry for the delay; I've been out of town)

**Reply by iheanyi** [reply]
> @iheanyi Yes, it does appear to be a problem with that specific database. I created a new database and this feature is working as expected there. In the malfunctioning database I've tried toggling the automatic migrations setting off and on and created several branches with no luck.
> 
> I'm happy to delete and recreate the database (I'd like to keep the name I chose for this database), but will wait for your word in case you want to investigate further.
> 
> (sorry for the delay; I've been out of town)

Yikes, I'd love to get the name of the organization and database so we can investigate it come Monday. Please send it to iheanyi@planetscale.com when you get a chance!

**Comment by rjvim** [comment]
Though the migration copy option was there I felt this should be handled a bit more manually using pscale cli and have control over it. I have detailed the problem and solution here: https://github.com/teurons/snippets/blob/master/pscale/migrate.md 

Hope it helps
